File: 0437-path-sum-iii/0437-path-sum-iii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func dfs(root *TreeNode, targetSum int, total int) int {
    if root == nil {
        return 0
    }

    total += root.Val
    if total == targetSum {
        return 1 + dfs(root.Left, targetSum, total) + dfs(root.Right, targetSum, total) 
    }

    return dfs(root.Left, targetSum, total) + dfs(root.Right, targetSum, total) 
}

func pathSum(root *TreeNode, targetSum int) int {
    if root == nil {
        return 0
    }

    return dfs(root, targetSum, 0) + pathSum(root.Left, targetSum) + pathSum(root.Right, targetSum)
}

File: 0437-path-sum-iii/0437-path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if not root:
            return 0

        def dfs(root, target):
            if not root:
                return 0

            return (target == root.val) + dfs(root.left, target - root.val) + dfs(root.right, target - root.val)
        
        return dfs(root, targetSum) + self.pathSum(root.left, targetSum) + self.pathSum(root.right, targetSum)
